main_test.go

```diff
@@ -510,6 +515,74 @@ var testCases = []TestCase{
 	},
 }
 
+type FakeS3Client struct {
+	Get mockGetObjectAPI
+	Put mockPutObjectAPI
+}
+
+type mockGetObjectAPI func(ctx context.Context, params *s3.GetObjectInput) (*s3.GetObjectOutput, error)
+type mockPutObjectAPI func(ctx context.Context, params *s3.PutObjectInput) (*s3.PutObjectOutput, error)
+
+func (fc *FakeS3Client) GetObject(ctx context.Context, params *s3.GetObjectInput, optFns ...func(*s3.Options)) (*s3.GetObjectOutput, error) {
+	return fc.Get(ctx, params)
+}
+
+func (fc *FakeS3Client) PutObject(ctx context.Context, params *s3.PutObjectInput, optFns ...func(*s3.Options)) (*s3.PutObjectOutput, error) {
+	return fc.Put(ctx, params)
+}
+
+func (fc *FakeS3Client) New(filename string) *FakeS3Client {
+	return &FakeS3Client{
```

Comment:
it looks like for `Get` this mock always reads the same file (regardless of the bucket/key specified), and for `Put` it writes to an arbitrary file based on the key (ignoring the bucket)? I think it would be a little cleaner if you did one of the two options:

1. Always reads and writes from/to the same file. Takes an expected bucket on creation and key and fails any request that doesn't match.
2. Reads and writes from/to the file at `/s3/<bucket>/<key>` (/s3 prefix makes sure it doesn't accidentally overwrite any existing files).

Either one of these options could catch some errors that the current code doesn't catch.

Option 1 seems simpler, so I would do that, unless you envision needing to read/write multiple files in the future.

Reply:
I do not need to read/write multiple files in the future. I now implement option 1, storing the file in /tmp/bucket/`in.xlsx` .